python_modules/libraries/dagster-openai/dagster_openai/resources.py

```diff
@@ -0,0 +1,254 @@
+    "chat": [["completions", "create"]],
+    "embeddings": [["create"]],
+    "files": [
+        ["create"],
+        ["retrieve"],
+        ["list"],
+        ["delete"],
+        ["content"],
+        ["retrieve_content"],
+        ["wait_for_processing"],
+    ],
+    "images": [["create_variation"], ["edit"], ["generate"]],
+    "audio": [["transcriptions", "create"], ["translations", "create"], ["speech", "create"]],
+    "moderations": [["create"]],
+    "models": [["retrieve"], ["list"], ["delete"]],
+    "fine_tuning": [
+        ["jobs", "create"],
+        ["jobs", "retrieve"],
+        ["jobs", "list"],
+        ["jobs", "cancel"],
+        ["jobs", "list_events"],
+    ],
+}
+
+context_to_counters = WeakKeyDictionary()
+
+
+def add_to_asset_metadata(
+    context: AssetExecutionContext, metadata_key: str, delta: int, output_name: Optional[str] = None
+):
+    if context not in context_to_counters:
+        context_to_counters[context] = defaultdict(lambda: 0)
+    counters = context_to_counters[context]
+    counters[metadata_key] += delta
+    context.add_output_metadata(dict(counters), output_name)
+
+
+def with_usage_metadata(context: AssetExecutionContext, output_name: Optional[str], func):
+    """This wrapped can be used on any endpoint of the
+    `openai library <https://github.com/openai/openai-python>`
+    to log the OpenAI API usage to asset metadata.
+
+    Examples:
+    .. code-block:: python
+
+        import os
+
+        from dagster import asset, define_asset_job, EnvVar
+        from dagster_openai import OpenAIResource, with_usage_metadata
+
+
+        @asset(compute_kind="OpenAI")
+        def openai_asset(context: AssetExecutionContext, openai: OpenAIResource):
+            with openai.get_client(context) as client:
+                client.fine_tuning.jobs.create = with_usage_metadata(
+                    client.fine_tuning.jobs.create
+                )
+                client.fine_tuning.jobs.create(
+                    model="gpt-3.5-turbo",
+                    training_file="some_training_file"
+                )
+
+        openai_asset_job = define_asset_job(name="openai_asset_job", selection="openai_asset")
+
+        defs = Definitions(
+            jobs=[openai_asset_job],
+            resources={
+                "slack": OpenAIResource(api_key=EnvVar("OPENAI_API_KEY")),
+            },
+        )
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        response = func(*args, **kwargs)
+        usage = response.usage
+        add_to_asset_metadata(context, "openai.calls", 1, output_name)
+        add_to_asset_metadata(context, "openai.total_tokens", usage.total_tokens, output_name)
+        add_to_asset_metadata(context, "openai.prompt_tokens", usage.prompt_tokens, output_name)
+        if hasattr(usage, "completion_tokens"):
+            add_to_asset_metadata(
+                context, "openai.completion_tokens", usage.completion_tokens, output_name
+            )
```

Comment:
nit: we can probably call the add_to_asset_metadata once since we will have all the dict value ready here [1]